pkg/dyninst/decode/decoder.go

```diff
@@ -0,0 +1,407 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package decode
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"reflect"
+	"strconv"
+
+	"github.com/go-json-experiment/json/jsontext"
+
+	"github.com/DataDog/datadog-agent/pkg/dyninst/ir"
+	"github.com/DataDog/datadog-agent/pkg/dyninst/output"
+)
+
+type probeEvent struct {
+	event *ir.Event
+	probe *ir.Probe
+}
+
+// Decoder decodes the output of the BPF program into a JSON format.
+type Decoder struct {
+	program                *ir.Program
+	typesToProbesAndEvents map[ir.TypeID]probeEvent
+}
+
+var errorUnimplemented = errors.New("errorUnimplemented type")
+
+// NewDecoder creates a new Decoder for the given program.
+func NewDecoder(program *ir.Program) (*Decoder, error) {
+	decoder := &Decoder{
+		program:                program,
+		typesToProbesAndEvents: make(map[ir.TypeID]probeEvent),
+	}
+	for _, probe := range program.Probes {
+		for _, event := range probe.Events {
+			decoder.typesToProbesAndEvents[event.Type.ID] = probeEvent{
+				event: event,
+				probe: probe,
+			}
+		}
+	}
+	return decoder, nil
+}
+
+type dataItemWithAddrCounter struct {
+	dataItem    output.DataItem
+	addrCounter uint8
+}
+
+// Decode decodes the given event into the given writer.
+func (d *Decoder) Decode(event output.Event, out io.Writer) error {
+	var (
+		rootData  []byte
+		eventRoot *ir.EventRootType
+		ok        bool
+		err       error
+		frames    []uint64
+	)
+
+	frames, err = event.StackPCs()
+	if err != nil {
+		return err
+	}
+	enc := jsontext.NewEncoder(out)
+	err = enc.WriteToken(jsontext.BeginObject)
+	if err != nil {
+		return err
+	}
+
+	err = enc.WriteToken(jsontext.String("stack_frames"))
+	if err != nil {
+		return err
+	}
+	err = enc.WriteToken(jsontext.BeginArray)
+	if err != nil {
+		return err
+	}
+	for i := range frames {
+		err = enc.WriteToken(jsontext.Uint(frames[i]))
+		if err != nil {
+			return err
+		}
+	}
+	err = enc.WriteToken(jsontext.EndArray)
+	if err != nil {
+		return err
+	}
+
+	err = enc.WriteToken(jsontext.String("captures"))
+	if err != nil {
+		return err
+	}
+	err = enc.WriteToken(jsontext.BeginObject)
+	if err != nil {
+		return err
+	}
+	itemsByAddress := map[uint64]*dataItemWithAddrCounter{}
+	for item, err := range event.DataItems() {
+		if errors.Is(err, output.ErrFinishedIterating) {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		if eventRoot == nil {
+			// item is EventRoot
```

Comment:
nit: does this comment add much value? I could see more exposition or just eliminating it.